toolchain/check/testdata/class/generic/self.carbon

```diff
@@ -0,0 +1,72 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// AUTOUPDATE
+// TIP: To test this file alone, run:
+// TIP:   bazel test //toolchain/testing:file_test --test_arg=--file_tests=toolchain/check/testdata/class/generic/self.carbon
+// TIP: To dump output, run:
+// TIP:   bazel run //toolchain/testing:file_test -- --dump_output --file_tests=toolchain/check/testdata/class/generic/self.carbon
+
+class Class(T:! type) {
+  // `Self` is the same as `Class(T)` here.
+  fn F[addr self: Self*]() -> Class(T)* {
+    return self;
+  }
```

Comment:
Test it works in both directions?
```suggestion
  fn F[addr self: Self*]() -> Class(T)* {
    return self;
  }
  fn G[addr self: Class(T)*]() -> Self* {
    return self;
  }
```

Reply:
Hm, I'm trying to test that it's the exact same type, not just that it's convertible. Maybe there's a better way to do that?

Reply:
I've changed this a bit to try to better test the types are the same. I'm still not really happy with it, but hopefully this will do for now.